### ISSUE TITLE
some testing and usability results

### ISSUE THREAD

**Issue by MatthiasGeorgImhof** [coderkalyan/libserard#2]

**Comment by pavel-kirienko** [comment]
>STM32 does not have bits/stdint-uintn.h. I had to exclude this include

Indeed, this is a compiler-specific header that must not be used. The correct header is `stdint.h`.

>serard.h misses some convenience typedefs, e.g. for

We ended up removing these from libudpard and libcanard, so this is fine.

>deviations from libcanard (which I used a lot for the past few weeks) should be minimized

This is an anti-goal. The transports are very different in their design so an attempt to minimize differences is likely to result in weird designs. It is not a bad idea to highlight commonalities, but these libraries will not look alike.

>struct SerardMemoryResource has userreference, deallocator, allocator. I would have expected allocator and deallocator switched since one typically talks about malloc and free, or allocator and deallocator

Seems like this was copied from libudpard, where this ordering is chosen to support structural subtyping. There is a [struct UdpardMemoryDeleter](https://github.com/OpenCyphal/libudpard/blob/3d9f5efc26b91d7c9bf9104f55551ea7ad454dc2/libudpard/udpard.h#L355-L359) that has no allocation function, hence the deallocation function must come first.

**Comment by coderkalyan** [comment]
Thanks for the feedback @MatthiasGeorgImhof!

> STM32 does not have bits/stdint-uintn.h. I had to exclude this include

Good catch. This must have been included by my language server, I will replace it with the correct `stdint.h`. 

> In serardRxAccept(), there is a test missing for out_subscription != NULL

Thanks, I'll add one. In general the RX pipeline test coverage is not complete (the TX pipeline testing is mostly complete).

> deviations from libcanard (which I used a lot for the past few weeks) should be minimized

In addition to what Pavel mentioned, libcanard is the oldest of the libraries and some API flaws have been discovered and fixed over time. You should find that libserard is much more similar to the libudpard API, although very simplified as it's a much lighter stack. `user_reference` is one such API difference that causes little to no overhead for C users but improves integration into other languages (C++, Rust, Zig).

> struct SerardMemoryResource has userreference, deallocator, allocator. I would have expected allocator and deallocator switched since one typically talks about malloc and free, or allocator and deallocator

Based on what Pavel mentioned above, I think this can be reordered. I don't use structural subtyping in this way for the allocator, although it is used extensively for the AVL tree, hence the careful ordering there.
